src/tpm2-tss-engine-ciphers.c

```diff
@@ -0,0 +1,393 @@
+};
+
+static int
+big_endian_array_to_native(const unsigned char *in, size_t size)
+{
+    uint32_t r;
+
+    memcpy(&r, in, size);
+    r = be32toh(r);
+
+    return r;
+}
+
+static int
+populate_tpm2data(const unsigned char *key, TPM2_DATA **tpm2Data)
+{
+    uint32_t keyHandle = 0;
+
+    /* Use an empty Key */
+    if (key == NULL) {
+        *tpm2Data = NULL;
+
+    /* Read persistent key, use for openssl app : openssl enc */
+    } else if ((key[0] == 0x81) && (key[HANDLE_SIZE/2 + 1] == 0)) {
+        keyHandle = big_endian_array_to_native(key, HANDLE_SIZE/2);
+        if (!tpm2tss_tpm2data_readtpm(keyHandle, tpm2Data))
+            return 0;
+
+    /* Use blob context */
+    } else {
+        return 0;
+    }
+
+    return 1;
+}
+
+static TPMI_ALG_SYM_MODE
+tpm2_get_cipher_mode(EVP_CIPHER_CTX *ctx, TPM2_DATA *tpm2Data)
+{
+    if (tpm2Data->pub.publicArea.parameters.symDetail.sym.mode.sym == TPM2_ALG_NULL) {
+        switch (EVP_CIPHER_CTX_mode(ctx)) {
+            case EVP_CIPH_CFB_MODE:
+                return TPM2_ALG_CFB;
+            case EVP_CIPH_OFB_MODE:
+                return TPM2_ALG_OFB;
+            case EVP_CIPH_CTR_MODE:
+                return TPM2_ALG_CTR;
+            case EVP_CIPH_ECB_MODE:
+                return TPM2_ALG_ECB;
+            case EVP_CIPH_CBC_MODE:
+                return TPM2_ALG_CBC;
+            default:
+                return TPM2_ALG_CBC;
+        }
+    } else {
+        return tpm2Data->pub.publicArea.parameters.symDetail.sym.mode.sym;
+    }
+}
+
+static int
+tpm2_cipher_init_key(EVP_CIPHER_CTX *ctx, const unsigned char *key, const unsigned char *iv, int enc)
+{
+    (void)(iv);
+    (void)(enc);
+    TPM2_DATA *tpm2Data = NULL;
+
+    DBG("Init Cipher Key ...\n");
+
+    /* Init Struct */
+    tpm2Data = OPENSSL_malloc(sizeof(TPM2_DATA));
+    if (tpm2Data == NULL) {
+        ERR(tpm2_cipher_init_key, ERR_R_MALLOC_FAILURE);
+        goto error;
+    }
+    memset(tpm2Data, 0, sizeof(TPM2_DATA));
+
+    /* Fill TPM2_DATA depending of KEY */
+    if (!populate_tpm2data(key, &tpm2Data)) {
+        ERR(tpm2_cipher_init_key, TPM2TSS_R_TPM2DATA_READ_FAILED);
+        goto error;
+    }
+
+    /* Set App Data */
+    EVP_CIPHER_CTX_set_app_data(ctx, tpm2Data);
+
+    DBG("algo : %#x", tpm2Data->pub.publicArea.parameters.symDetail.sym.algorithm);
+    DBG(" | mode : %#x", tpm2_get_cipher_mode(ctx, tpm2Data));
+    DBG(" | size : %d", tpm2Data->pub.publicArea.parameters.symDetail.sym.keyBits.sym);
+    DBG(" | enc  : %d", !EVP_CIPHER_CTX_encrypting(ctx));
+    DBG(" | iv   : %d\n", EVP_CIPHER_CTX_iv_length(ctx));
+
+    return 1;
+
+error :
+    if (tpm2Data)
+        OPENSSL_free(tpm2Data);
+    return 0;
+}
+
+static int
+tpm2_do_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out, const unsigned char *in, size_t inl)
+{
+    TPM2_DATA *tpm2Data;
+    TSS2_RC ret;
+    ESYS_AUXCONTEXT eactx = (ESYS_AUXCONTEXT){ NULL, NULL};
+    ESYS_TR keyHandle = ESYS_TR_NONE;
+    TPM2B_MAX_BUFFER *out_data, *in_data;
```

Comment:
in_data is never freed.
Also you don't have to malloc it, you could just declare `TPM2_MAX_BUFFER in_data;` on stack instead of a pointer.